HACKING.md

```diff
@@ -232,6 +232,121 @@ back the actions in the browser window you have open, so doesn't require any
 headless mode. However, before running a test in this way, you must make sure
 the test database is clean and has t/sql/selenium.sql loaded into it.
 
+### Schema change
+
+Prior to a schema change, you'll want to test that [upgrade.sh](upgrade.sh)
+runs without errors on the previous schema version.
+Our [sample database dump](http://ftp.musicbrainz.org/pub/musicbrainz/data/sample/)
+is useful for testing this on a standalone database, but it's also important
+to test against a full backup of the production database.
+
+Given its size, a full production test can be time consuming, particularly if
+the upgrade fails and you have to restore the dump again. To help ensure any
+failures are *data-related* and not *schema-related*, it's recommended to
+dump the production schema (without data) locally, and run an upgrade against
+that first.
+
+To do this, first create an empty "musicbrainz_prod_schema" database with the
+proper roles and privileges. Configure this as `PROD_SCHEMA_CHANGE_TEST`, or
+some other name of your choice, in DBDefs.pm.
+
+```sh
+# You may require custom host/port connection parameters to your local DB.
+psql -U postgres -d template1 -c 'CREATE DATABASE musicbrainz_prod_schema;'
+# It's okay if these already exists.
+psql -U postgres -d template1 -c 'CREATE ROLE musicbrainz;'
+psql -U postgres -d template1 -c 'CREATE ROLE musicbrainz_ro;'
+psql -U postgres -d template1 -c 'CREATE ROLE caa_redirect;'
+psql -U postgres -d template1 -c 'CREATE ROLE sir;'
+psql -U postgres -d template1 -c 'GRANT CREATE ON DATABASE musicbrainz_prod_schema TO musicbrainz;'
+```
+
+Next, you can dump the schema of the production standby, import it locally to
+`musicbrainz_prod_schema`, and run upgrade.sh against
+`PROD_SCHEMA_CHANGE_TEST`:
+
+```sh
+ssh pink docker exec postgres-pink sudo -E -H -u postgres \
+  pg_dump --schema-only musicbrainz_db \
+  > musicbrainz_prod_schema.dump
```

Comment:
I would suggest to:

- refer to the type of Postgres instance instead of the current host which is changing over time,
- compress the dump for transmission over SSH (`gzip` is the default with `ssh -C`),

as follows:

```suggestion
prod_standby_pg_host=
prod_standby_pg_container=
ssh -C $prod_standby_pg_host docker exec $prod_standby_pg_container sudo -E -H -u postgres \
  pg_dump --schema-only musicbrainz_db \
  | gunzip \
  > musicbrainz_prod_schema.dump
```

Reply:
I tested the new command locally and couldn't get the compression to work; gunzip would give `gzip: stdin: not in gzip format`, and running just the ssh command (including `-C`) did not return compressed output for me.

I did copy the other improvements you made, though.

Reply:
Oops, the `gunzip` command is unneeded as ssh is uncompressing it already.

Reply:
Oh, okay.  I'll just add the `-C` in and push that to master then. :)

Reply:
For the record, the commit is https://github.com/metabrainz/musicbrainz-server/commit/2cd0f55bcec4ebd3839b488fb379b1be3d52b69e.